### ISSUE TITLE
(@aws-cdk/custom-resources): don't format parameters as json string unless necessary to improve 'cdk diff'ability

### ISSUE THREAD

**Issue by heikkis** [aws/aws-cdk#13693]
### The Question

I would like to create AWSCustomResource with complex "JSON" string parameter. The formatting parameter changed after upgrading to 1.91.0 -> 1.92.0+ (until 1.94.1 at the moment).

The new format *is valid* (after all) but it is hard to read and understand anymore. 

I can see only this kind of JSON string as a parameter with escapes of escapes and starting from "Create" attribute:
```
        "Create": "{\"service\":\"Organizations\",\"action\":\"createPolicy\",\"parameters\":{\"Content\":\"{\\\"Version\\\":\\\"2012-10-17\\\",\\\"Statement\\\":[{\\\"Sid\\\":\\\"ProtectOrganization\\\",\\\"Effect\\\":\\\"Deny\\\",\\\"Action\\\":[\\\"organizations:*\\\"],\\\"Resource\\\":\\\"*\\\"}]}\",\"Description\":\"Desc\",\"Name\":\"GITHUB-13693\",\"Type\":\"SERVICE_CONTROL_POLICY\"},\"outputPath\":\"Policy.PolicySummary\",\"physicalResourceId\":{\"responsePath\":\"Policy.PolicySummary.Id\"}}",
```

Compared the older one up to 1.91.0:
```
        "Create": {
          "service": "Organizations",
          "action": "createPolicy",
          "parameters": {
            "Content": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"ProtectOrganization\",\"Effect\":\"Deny\",\"Action\":[\"organizations:*\"],\"Resource\":\"*\"}]}",
            "Description": "Desc",
            "Name": "GITHUB-13693",
            "Type": "SERVICE_CONTROL_POLICY"
          },
          "outputPath": "Policy.PolicySummary",
          "physicalResourceId": {
            "responsePath": "Policy.PolicySummary.Id"
          }
        },
```

More details in a comment below: https://github.com/aws/aws-cdk/issues/13693#issuecomment-803297453

<details>
<summary>Example code</summary>

```
import cdk = require('@aws-cdk/core');
import {Construct, Stack, StackProps} from "@aws-cdk/core";
import {AwsCustomResource, AwsCustomResourcePolicy, PhysicalResourceId} from "@aws-cdk/custom-resources";
import {DefaultAdminPrincipals} from "./shared/config/scp_principals";

export const APP = new cdk.App();

export class TestStack extends Stack {
    private getStatementsAsPolicyJSON() {
        return ({
            'Version': '2012-10-17',
            'Statement': [
                {
                    Sid: 'ProtectOrganization',
                    Effect: 'Deny',
                    Action: ['organizations:*'],
                    Resource: '*'
                }
            ]
        });
    }

    constructor(scope: Construct, id: string, props: StackProps) {
        super(scope, id, props);

        new AwsCustomResource(this, 'GITHUB-13693', {
            onCreate: {
                service: 'Organizations',
                action: 'createPolicy',
                parameters: {
                    Content: JSON.stringify(this.getStatementsAsPolicyJSON()),
                    Description: 'Desc',
                    Name: 'GITHUB-13693',
                    Type: 'SERVICE_CONTROL_POLICY',
                },
                outputPath: 'Policy.PolicySummary',
                physicalResourceId: PhysicalResourceId.fromResponse('Policy.PolicySummary.Id'),
            },
            policy: AwsCustomResourcePolicy.fromSdkCalls({
                resources: AwsCustomResourcePolicy.ANY_RESOURCE,
            }),
            installLatestAwsSdk: false,
        });
    }
}

new TestStack(APP, 'GITHUB-13693', {
    env: {
        region: 'us-east-1',
        account: '123456789012345',
    }
})

APP.synth()

```

</details>

### Environment

  - **CDK CLI Version:  1.92.0 (build de536cc)** 
  - **Module Version: 1.92.0** 
  - **Node.js Version: v13.10.1** 
  - **OS: MacOS latest** 
  - **Language (Version):TypeScript (4.2.3) ** 

### Other information
Google offered this issue https://github.com/aws/aws-cdk/pull/13074 but I could not find out the answer.

**Comment by jogold** [comment]
What is the error you're seeing when deploying?

**Comment by heikkis** [comment]
Thanks @jogold for reply. The original topic was about error in produced attribute, but that was my mistake when trying to avoid escape of escapes in CF template.

So my question is about the formatting which changed after 1.92.0 to ambiguous IMO. 

I can see only this kind of JSON string as a parameter with escapes of escapes and starting from "Create" attribute:
```
        "Create": "{\"service\":\"Organizations\",\"action\":\"createPolicy\",\"parameters\":{\"Content\":\"{\\\"Version\\\":\\\"2012-10-17\\\",\\\"Statement\\\":[{\\\"Sid\\\":\\\"ProtectOrganization\\\",\\\"Effect\\\":\\\"Deny\\\",\\\"Action\\\":[\\\"organizations:*\\\"],\\\"Resource\\\":\\\"*\\\"}]}\",\"Description\":\"Desc\",\"Name\":\"GITHUB-13693\",\"Type\":\"SERVICE_CONTROL_POLICY\"},\"outputPath\":\"Policy.PolicySummary\",\"physicalResourceId\":{\"responsePath\":\"Policy.PolicySummary.Id\"}}",
```

Compared the older one up to 1.91.0:
```
        "Create": {
          "service": "Organizations",
          "action": "createPolicy",
          "parameters": {
            "Content": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"ProtectOrganization\",\"Effect\":\"Deny\",\"Action\":[\"organizations:*\"],\"Resource\":\"*\"}]}",
            "Description": "Desc",
            "Name": "GITHUB-13693",
            "Type": "SERVICE_CONTROL_POLICY"
          },
          "outputPath": "Policy.PolicySummary",
          "physicalResourceId": {
            "responsePath": "Policy.PolicySummary.Id"
          }
        },
```



Full CF template with CDK 1.92.0:
```
{
  "Resources": {
    "GITHUB13693CustomResourcePolicyD78D7C4B": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "organizations:CreatePolicy",
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "GITHUB13693CustomResourcePolicyD78D7C4B",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2"
          }
        ]
      },
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/GITHUB-13693/CustomResourcePolicy/Resource"
      }
    },
    "GITHUB136938EEF37A0": {
      "Type": "Custom::AWS",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn"
          ]
        },
        "Create": "{\"service\":\"Organizations\",\"action\":\"createPolicy\",\"parameters\":{\"Content\":\"{\\\"Version\\\":\\\"2012-10-17\\\",\\\"Statement\\\":[{\\\"Sid\\\":\\\"ProtectOrganization\\\",\\\"Effect\\\":\\\"Deny\\\",\\\"Action\\\":[\\\"organizations:*\\\"],\\\"Resource\\\":\\\"*\\\"}]}\",\"Description\":\"Desc\",\"Name\":\"GITHUB-13693\",\"Type\":\"SERVICE_CONTROL_POLICY\"},\"outputPath\":\"Policy.PolicySummary\",\"physicalResourceId\":{\"responsePath\":\"Policy.PolicySummary.Id\"}}",
        "InstallLatestAwsSdk": false
      },
      "DependsOn": [
        "GITHUB13693CustomResourcePolicyD78D7C4B"
      ],
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete",
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/GITHUB-13693/Resource/Default"
      }
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      },
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/AWS679f53fac002430cb0da5b7982bd2287/ServiceRole/Resource"
      }
    },
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-fintraffic-assets-195932357633-us-east-1",
          "S3Key": "a5a55f40f05c1031f30cc935c76b58b0b7dbc81e70845edf4ef615d6580ae04d.zip"
        },
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn"
          ]
        },
        "Handler": "index.handler",
        "Runtime": "nodejs12.x",
        "Timeout": 120
      },
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2"
      ],
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/AWS679f53fac002430cb0da5b7982bd2287/Resource",
        "aws:asset:path": "asset.a5a55f40f05c1031f30cc935c76b58b0b7dbc81e70845edf4ef615d6580ae04d",
        "aws:asset:property": "Code"
      }
    },
    "CDKMetadata": {
      "Type": "AWS::CDK::Metadata",
      "Properties": {
        "Modules": "aws-cdk=1.92.0,@aws-cdk/assets=1.92.0,@aws-cdk/aws-applicationautoscaling=1.92.0,@aws-cdk/aws-autoscaling-common=1.92.0,@aws-cdk/aws-cloudwatch=1.92.0,@aws-cdk/aws-codeguruprofiler=1.92.0,@aws-cdk/aws-ec2=1.92.0,@aws-cdk/aws-ecr=1.92.0,@aws-cdk/aws-ecr-assets=1.92.0,@aws-cdk/aws-events=1.92.0,@aws-cdk/aws-iam=1.92.0,@aws-cdk/aws-kms=1.92.0,@aws-cdk/aws-lambda=1.92.0,@aws-cdk/aws-logs=1.92.0,@aws-cdk/aws-s3=1.92.0,@aws-cdk/aws-s3-assets=1.92.0,@aws-cdk/aws-sqs=1.92.0,@aws-cdk/aws-ssm=1.92.0,@aws-cdk/cloud-assembly-schema=1.92.0,@aws-cdk/core=1.92.0,@aws-cdk/custom-resources=1.92.0,@aws-cdk/cx-api=1.92.0,@aws-cdk/region-info=1.92.0,jsii-runtime=node.js/v13.10.1"
      },
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/CDKMetadata/Default"
      }
    }
  },
  "Parameters": {
    "BootstrapVersion": {
      "Type": "AWS::SSM::Parameter::Value<String>",
      "Default": "/cdk-bootstrap/fintraffic/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store."
    }
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3"
                  ],
                  {
                    "Ref": "BootstrapVersion"
                  }
                ]
              }
            ]
          },
          "AssertDescription": "CDK bootstrap stack version 4 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI."
        }
      ]
    }
  }
}
```

Full CF template with CDK 1.91.0:
```
{
  "Resources": {
    "GITHUB13693CustomResourcePolicyD78D7C4B": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "organizations:CreatePolicy",
              "Effect": "Allow",
              "Resource": "*"
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "GITHUB13693CustomResourcePolicyD78D7C4B",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2"
          }
        ]
      },
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/GITHUB-13693/CustomResourcePolicy/Resource"
      }
    },
    "GITHUB136938EEF37A0": {
      "Type": "Custom::AWS",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn"
          ]
        },
        "Create": {
          "service": "Organizations",
          "action": "createPolicy",
          "parameters": {
            "Content": "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Sid\":\"ProtectOrganization\",\"Effect\":\"Deny\",\"Action\":[\"organizations:*\"],\"Resource\":\"*\"}]}",
            "Description": "Desc",
            "Name": "GITHUB-13693",
            "Type": "SERVICE_CONTROL_POLICY"
          },
          "outputPath": "Policy.PolicySummary",
          "physicalResourceId": {
            "responsePath": "Policy.PolicySummary.Id"
          }
        },
        "InstallLatestAwsSdk": false
      },
      "DependsOn": [
        "GITHUB13693CustomResourcePolicyD78D7C4B"
      ],
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete",
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/GITHUB-13693/Resource/Default"
      }
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      },
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/AWS679f53fac002430cb0da5b7982bd2287/ServiceRole/Resource"
      }
    },
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-fintraffic-assets-195932357633-us-east-1",
          "S3Key": "4a3609ad912843e581892f37ae9d6fb0fa1648b547693aaa562b0119452b8956.zip"
        },
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn"
          ]
        },
        "Handler": "index.handler",
        "Runtime": "nodejs12.x",
        "Timeout": 120
      },
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2"
      ],
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/AWS679f53fac002430cb0da5b7982bd2287/Resource",
        "aws:asset:path": "asset.4a3609ad912843e581892f37ae9d6fb0fa1648b547693aaa562b0119452b8956",
        "aws:asset:property": "Code"
      }
    },
    "CDKMetadata": {
      "Type": "AWS::CDK::Metadata",
      "Properties": {
        "Modules": "aws-cdk=1.91.0,@aws-cdk/assets=1.91.0,@aws-cdk/aws-applicationautoscaling=1.91.0,@aws-cdk/aws-autoscaling-common=1.91.0,@aws-cdk/aws-cloudwatch=1.91.0,@aws-cdk/aws-codeguruprofiler=1.91.0,@aws-cdk/aws-ec2=1.91.0,@aws-cdk/aws-ecr=1.91.0,@aws-cdk/aws-ecr-assets=1.91.0,@aws-cdk/aws-events=1.91.0,@aws-cdk/aws-iam=1.91.0,@aws-cdk/aws-kms=1.91.0,@aws-cdk/aws-lambda=1.91.0,@aws-cdk/aws-logs=1.91.0,@aws-cdk/aws-s3=1.91.0,@aws-cdk/aws-s3-assets=1.91.0,@aws-cdk/aws-sqs=1.91.0,@aws-cdk/aws-ssm=1.91.0,@aws-cdk/cloud-assembly-schema=1.91.0,@aws-cdk/core=1.91.0,@aws-cdk/custom-resources=1.91.0,@aws-cdk/cx-api=1.91.0,@aws-cdk/region-info=1.91.0,jsii-runtime=node.js/v13.10.1"
      },
      "Metadata": {
        "aws:cdk:path": "GITHUB-13693/CDKMetadata/Default"
      }
    }
  },
  "Parameters": {
    "BootstrapVersion": {
      "Type": "AWS::SSM::Parameter::Value<String>",
      "Default": "/cdk-bootstrap/fintraffic/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store."
    }
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3"
                  ],
                  {
                    "Ref": "BootstrapVersion"
                  }
                ]
              }
            ]
          },
          "AssertDescription": "CDK bootstrap stack version 4 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI."
        }
      ]
    }
  }
}
```

**Comment by rix0rrr** [comment]
So there is no bug, the issue is that you USED to be able to read (human read?) the JSON, more or less, and now you can't easily anymore, right?

The change in question was introduced here to fix another issue: https://github.com/aws/aws-cdk/pull/13074

There is unfortunately no way around it. There is no public contract on the template part that is produced by this custom resource, and even if there were, reading it comfortably as a human would probably not be part of that contract.

If you must, you can write a script to read and decode that part of the template.

But really, I'm more interested in why you WOULD. Why do you care what the template looks like?

**Comment by heikkis** [comment]
Yep, I understand that there is no contract for cloud formation formation.

When you are developing changes to CDK project, I like to see clear cdk diff output to analyze that what I have changed and what I have not, even with custom resources. Before I got see that the change is about eg. content, physical id or another parameter, and now after 1.92.0 it is all one big json string and cdk diff tool does not make any sense of it anymore.
